### PR TITLE
K8s io update to 24.0

### LINKED ISSUE #5009: Update Kubernetes dependency to 0.24.0
**Describe the problem/challenge you have**
When updating k8s.io to v0.24.0 in my project, I get failures in velero cannot update
Example Error:
```
github.com/vmware-tanzu/velero/pkg/builder
# github.com/vmware-tanzu/velero/pkg/builder
../../go/pkg/mod/github.com/vmware-tanzu/velero@v1.8.1/pkg/builder/object_meta.go:124:7: obj.SetClusterName undefined (type "k8s.io/apimachinery/pkg/apis/meta/v1".Object has no field or method SetClusterName)
make: *** [build-real] Error 2
```

**Describe the solution you'd like**
Update k8s.io components from v0.22.2 to v0.24.0

**Environment:**
N/A

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

### PR DESCRIPTION
**Pull request by kcboyle**
# Please add a summary of your change
Bump k8s.io dependencies for compatibility with packages that use velero but have updated this version of k8s.

# Does your change fix a particular issue?

Fixes #5009

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [x] Updated the corresponding documentation in `site/content/docs/main`.

- Not updating documentation, as this is a dependency bump

### CODE REVIEW

**Comment by kcboyle** [comment]
`make test` was run against the branch with these changes with success. Is there anything I could do to help this process go more smoothly? Thank you

**Comment by laverya** [comment]
Just like with #2651, other projects cannot import both velero and packages relying on k8s 1.24+ without a change like this being merged.

This one's a lot smaller though 😆

File: pkg/restore/restore_test.go [hunk 1]
```diff
@@ -871,7 +871,6 @@ func TestRestoreItems(t *testing.T) {
 						ObjectMeta(
 							builder.WithLabels("key-1", "val-1"),
 							builder.WithAnnotations("key-1", "val-1"),
-							builder.WithClusterName("cluster-1"),
```
Comment (blackpiglet) [inline pkg/restore/restore_test.go:874]: This change is because of ClusterName is deprecated in v1.24 metadata, right?

**Review by blackpiglet — DISMISSED** [review]

**Review by blackpiglet — COMMENTED** [review]
There are two check failures. Please clear the errors.

**Comment by blackpiglet** [comment]
@kcboyle 
If you don't have enough time to continue working on this PR, would you mind I create a new one to finish this task?

**Comment by laverya** [comment]
@blackpiglet to the best of my knowledge, upgrading k8s libraries will not impact compatibility with previous k8s versions - just like you can use 1.24.x kubectl with 1.16.x kubernetes clusters.

Unless there's resource types that have been completely removed that are being used, and that would cause compile errors instead.

**Comment by kcboyle** [comment]
I get an error when trying to run `make update` and fix one of the check errors:
```
make update
Using Cached Image: velero/build-image:3bb6252d
Running all update scripts
Updating gofmt
Updating gofmt - done!
Updating goimports
Updating goimports - done!
+ [[ -z /go ]]
+ [[ ! -d /go/src/k8s.io/code-generator ]]
+ command -v controller-gen
+ /go/src/k8s.io/code-generator/generate-groups.sh all github.com/vmware-tanzu/velero/pkg/generated github.com/vmware-tanzu/velero/pkg/apis velero:v1 --go-header-file ./hack/boilerplate.go.txt --output-base ../../..
Generating deepcopy funcs
Generating clientset for velero:v1 at github.com/vmware-tanzu/velero/pkg/generated/clientset
Generating listers for velero:v1 at github.com/vmware-tanzu/velero/pkg/generated/listers
Generating informers for velero:v1 at github.com/vmware-tanzu/velero/pkg/generated/informers
+ controller-gen crd:crdVersions=v1 paths=./pkg/apis/velero/v1/... rbac:roleName=velero-perms paths=./pkg/controller/... output:crd:artifacts:config=config/crd/v1/bases object paths=./pkg/apis/velero/v1/...
++ cat hack/restore-crd-patch-v1.json
+ kubectl patch -f config/crd/v1/bases/velero.io_restores.yaml -p '[
    { "op": "replace", "path": "/spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/hooks/properties/resources/items/properties/postHooks/items/properties/init/properties/initContainers/items/properties/ports/items/required", "value": [ "containerPort", "protocol"] }
]' --type=json --local=true -o yaml
+ mv /tmp/velero.io_restores-yaml.patched config/crd/v1/bases/velero.io_restores.yaml
+ go generate ./config/crd/v1/crds
Updating generated Github issue template
Success!
Updating plugin proto
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
make[1]: *** [shell] Error 255
make: *** [update] Error 2
```

I'm sure I missed a step somewhere in the setup/testing. Could you please advise?

**Comment by kcboyle** [comment]
Also, looking into the other failure, it looks like it's seeing this go sum as having a misspelling of "yet" ("EyT") 
```
go.uber.org/goleak v1.1.12 h1:gZAh5/EyT/HQwlpkCy6wTpqfH9H8Lz8zbm3dZh+OyzA=
```

It seems like the workflow should ignore spellings in go mod and go sum files, as there's often random strings of letters like this in those files. I made a commit to add them to the skip list. Please let me know if there is a better course of action here

**Comment by blackpiglet** [comment]
@kcboyle 
Thanks. Two Github actions should be modified. I will create a PR to fix it.

**Comment by blackpiglet** [comment]
@kcboyle 
I didn't some verification on my fork repo, and it worked. Need to merge main branch change, not release-1.9. IMO, need to have recently merged Github action change to work.

**Comment by blackpiglet** [comment]
@kcboyle 
Spell and CRD checks passed, but still needs to run `make update` to apply the CRDs change.
Sorry for the inconvenience in contributing, we will keep the actions update later.

**Comment by kcboyle** [comment]
re-ran the `make update` and re-pushed the result :)

**Review by blackpiglet — APPROVED** [review]

**Comment by blackpiglet** [comment]
@kcboyle 
Finally passed! Thanks for contribution.

**Comment by anshulahuja98** [comment]
@blackpiglet  / @kcboyle by when will this get merged?
And next release ETA?

**Comment by blackpiglet** [comment]
@anshulahuja98 
I think this PR is OK to merge, but Velero team still thinks there may be some impact on older versions of k8s. At least there should be some work for verifying Velero still works on v1.16 k8s cluster after this change, but Velero team doesn't have the environment for test.
https://github.com/vmware-tanzu/velero/issues/5009#issuecomment-1161393344

**Comment by ShimiT** [comment]
@blackpiglet - any update on this? as the current version use fairly old k8s lib's

**Comment by blackpiglet** [comment]
> @blackpiglet - any update on this? as the current version use fairly old k8s lib's

@ShimiT 
By far, no update yet, because Velero team doesn't have a v1.16 k8s cluster for verification.
By the way, may I ask why your case needs update Velero k8s liberary version?

**Comment by sunglim** [comment]
Hi 

Could we make this PR submitted?  My personal project imports velero package. but it prints an error while building it.

I really need a new version including this PR.

```
 > [builder 10/10] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go:
#17 23.15 # github.com/vmware-tanzu/velero/pkg/builder
#17 23.15 /go/pkg/mod/github.com/vmware-tanzu/velero@v1.7.0/pkg/builder/object_meta.go:124:7: obj.SetClusterName undefined (type "k8s.io/apimachinery/pkg/apis/meta/v1".Object has no field or method SetClusterName)
```

**Comment by d1egoaz** [comment]
> Hi
> 
> Could we make this PR submitted? My personal project imports velero package. but it prints an error while building it.
> 
> I really need a new version including this PR.
> 
> ```
>  > [builder 10/10] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go:
> #17 23.15 # github.com/vmware-tanzu/velero/pkg/builder
> #17 23.15 /go/pkg/mod/github.com/vmware-tanzu/velero@v1.7.0/pkg/builder/object_meta.go:124:7: obj.SetClusterName undefined (type "k8s.io/apimachinery/pkg/apis/meta/v1".Object has no field or method SetClusterName)
> ```

if you are ok with this branch's code, you can do this in your `go.mod` file until this gets merged:

```
// https://github.com/vmware-tanzu/velero/pull/5012
replace github.com/vmware-tanzu/velero => github.com/kcboyle/velero v1.7.0-rc.1.0.20220630140322-e5bb9459a4f3
```

This will make the dependency follow this commit https://github.com/vmware-tanzu/velero/pull/5012/commits/e5bb9459a4f3b228eb13c14cf40613e76a52c0b1

**Comment by reasonerjt** [comment]
@kcboyle would you have time to rebase this PR?

**Comment by ywk253100** [comment]
@kcboyle As we are very close to the FC date, I submitted another [PR](https://github.com/vmware-tanzu/velero/pull/5471) to address this issue. 

Your commit is included in the new PR. Thanks for your contribution!

I'm closing this one.

**Pull request closed without merging**